lib/DxilPIXPasses/DxilPIXAddTidToAmplificationShaderPayload.cpp

```diff
@@ -45,143 +45,154 @@ void DxilPIXAddTidToAmplificationShaderPayload::applyOptions(PassOptions O) {
 }
 
 void AddValueToExpandedPayload(OP *HlslOP, llvm::IRBuilder<> &B,
-                               ExpandedStruct &expanded,
                                AllocaInst *NewStructAlloca,
                                unsigned int expandedValueIndex, Value *value) {
   Constant *Zero32Arg = HlslOP->GetU32Const(0);
   SmallVector<Value *, 2> IndexToAppendedValue;
   IndexToAppendedValue.push_back(Zero32Arg);
   IndexToAppendedValue.push_back(HlslOP->GetU32Const(expandedValueIndex));
   auto *PointerToEmbeddedNewValue = B.CreateInBoundsGEP(
-      expanded.ExpandedPayloadStructType, NewStructAlloca, IndexToAppendedValue,
+      NewStructAlloca, IndexToAppendedValue,
       "PointerToEmbeddedNewValue" + std::to_string(expandedValueIndex));
   B.CreateStore(value, PointerToEmbeddedNewValue);
 }
 
-bool DxilPIXAddTidToAmplificationShaderPayload::runOnModule(Module &M) {
+void CopyAggregate(IRBuilder<> &B, Type *Ty, Value *Source, Value *Dest,
+                   ArrayRef<Value *> GEPIndices) {
+  if (StructType *ST = dyn_cast<StructType>(Ty)) {
+    SmallVector<Value *, 16> StructIndices;
+    StructIndices.append(GEPIndices.begin(), GEPIndices.end());
+    StructIndices.push_back(nullptr);
+    for (unsigned j = 0; j < ST->getNumElements(); ++j) {
+      StructIndices.back() = B.getInt32(j);
+      CopyAggregate(B, ST->getElementType(j), Source, Dest, StructIndices);
+    }
+  } else if (ArrayType *AT = dyn_cast<ArrayType>(Ty)) {
+    SmallVector<Value *, 16> StructIndices;
+    StructIndices.append(GEPIndices.begin(), GEPIndices.end());
+    StructIndices.push_back(nullptr);
+    for (unsigned j = 0; j < AT->getNumElements(); ++j) {
+      StructIndices.back() = B.getInt32(j);
+      CopyAggregate(B, AT->getArrayElementType(), Source, Dest, StructIndices);
+    }
+  } else {
+    auto *SourceGEP = B.CreateGEP(Source, GEPIndices, "CopyStructSourceGEP");
+    Value *Val = B.CreateLoad(SourceGEP, "CopyStructLoad");
+    auto *DestGEP = B.CreateGEP(Dest, GEPIndices, "CopyStructDestGEP");
+    B.CreateStore(Val, DestGEP, "CopyStructStore");
+  }
+}
```

Comment:
I could swear I've seen something like this before, but the closest I came was the memcpy splitting code, which isn't quite right. Was this modified from anything existing?

Reply:
No, it was coded fresh.

Reply:
(The nearest thing I've seen descends a DIType hierarchy. Didn't find an example of iterating through a real type like this)